src/vim.ts

```diff
@@ -24,20 +26,104 @@ class Vim {
+  meshIdToNodeIndex: Map<number, [number]>
 
   constructor (
-    vim: Vim,
-    meshes: InstancedMesh<BufferGeometry, Material>[],
-    sphere: Sphere
+    meshes: Mesh[],
+    boundingSphere: THREE.Sphere,
+    nodeIndexToMeshInstance: Map<number, [Mesh, number]>,
+    meshIdToNodeIndex: Map<number, [number]>
   ) {
-    this.vim = vim
     this.meshes = meshes
-    this.boundingSphere = sphere
+    this.boundingSphere = boundingSphere
+    this.nodeIndexToMeshInstance = nodeIndexToMeshInstance
+    this.meshIdToNodeIndex = meshIdToNodeIndex
+  }
+}
+
+class VimScene {
+  vim: Vim
+  geometry: VimSceneGeometry
+  elementToNodes: Map<number, number[]>
+
+  constructor (vim: Vim, geometry: VimSceneGeometry) {
+    this.vim = vim
+    this.geometry = geometry
+    this.elementToNodes = this.mapElements()
+  }
+
+  private mapElements (): Map<number, number[]> {
+    const nodeCount = this.getNodeCount()
+    if (nodeCount <= 0) return null
+
+    const map = new Map<number, number[]>()
+    for (let i = 0; i < nodeCount; i++) {
+      const elementId = this.getElementIdFromNodeIndex(i)
+      if (map.has(elementId)) {
+        map.get(elementId).push(i)
+      } else {
+        map.set(elementId, [i])
+      }
+    }
+    return map
+  }
+
+  getNodeCount (): number {
+    return this.geometry.nodeIndexToMeshInstance.size
+  }
+
+  getNodeIndicesFromElement (elementId: number): number[] | null {
+    return this.elementToNodes?.has(elementId)
+      ? this.elementToNodes.get(elementId)
+      : null
+  }
+
+  getMeshesFromElement (elementId: number): number[] | [Mesh, number][] {
+    const result: [Mesh, number][] = []
+    const nodeIndices = this.getNodeIndicesFromElement(elementId)
+    nodeIndices.forEach((i) => {
+      result.push(this.getMeshFromNodeIndex(i))
+    })
+    return result
+  }
+
+  getMeshFromNodeIndex (nodeIndex: number): [Mesh, number] {
+    return this.geometry.nodeIndexToMeshInstance.get(nodeIndex)
+  }
+
+  getNodeIndexFromMesh (mesh: THREE.Mesh, instance: number): number | null {
+    return this.geometry.meshIdToNodeIndex.get(mesh.id)[instance]
+  }
+
+  getElementIdFromMesh (mesh: THREE.Mesh, instance: number): number | null {
+    const nodeIndex = this.getNodeIndexFromMesh(mesh, instance)
+    return nodeIndex ? this.getElementIndexFromNodeIndex(nodeIndex) : null
+  }
+
+  getElementIndexFromNodeIndex (nodeIndex: number) {
+    return this.vim.bim.get('Vim.Node').get('Rvt.Element')[nodeIndex]
+  }
+
+  getElementIdFromNodeIndex (nodeIndex: number) {
+    const elementIndex = this.getElementIndexFromNodeIndex(nodeIndex)
+    const elementId = this.vim.bim.get('Rvt.Element').get('Id')[elementIndex]
+    return elementId
+  }
+
+  // TODO add better ways to access bim
+  getElementNameFromNodeIndex (nodeIndex: number) {
+    const elementIndex = this.getElementIndexFromNodeIndex(nodeIndex)
+    const nameIndex = this.vim.bim.get('Rvt.Element').get('Name')[elementIndex]
```

Comment:
This is a very slow lookup. I would create an array of all element names in the constructor of the Vim scene.

Reply:
Seems like early optimization to me. Both are O(1)